src/LanguageServer/Impl/Diagnostics/DiagnosticsService.cs

```diff
@@ -74,6 +83,16 @@ internal sealed class DiagnosticsService : IDiagnosticsService, IDisposable {
         }
 
         public int PublishingDelay { get; set; } = 1000;
+
+        public DiagnosticsSeverityMap DiagnosticsSeverityMap {
+            get => _severityMap;
+            set {
+                lock (_lock) {
+                    _severityMap = value;
+                    PublishDiagnostics();
```

Comment:
One more place under lock.

Reply:
As a matter of fact, I made PublishDiagnostics inside locks intentionally. The problem here if publishing goes off the lock, then we can go out of sync between `_changed` and the state of the collection. For example,

```cs
        public void Remove(Uri documentUri) {
            lock (_lock) {
                // Before removing the document, make sure we clear its diagnostics.
                _diagnostics[documentUri] = new List<DiagnosticsEntry>();
                PublishDiagnostics();
                _diagnostics.Remove(documentUri);
            }
        }
```

`PublishDiagnostics` must be called between `_diagnostics[documentUri] = new List<DiagnosticsEntry>()` which makes collection of the error empty and VSC clears the task list and `_diagnostics.Remove(documentUri)` which removes the document. Calling `PublishDiagnostics` after removal will lead to stale errors in the error list.